### PR TITLE
Translate cast expression

### PR DESCRIPTION
**Pull request by itssamuelrowe**

### CODE REVIEW

File: pallene/translator.lua [hunk 1]
```diff
@@ -85,6 +85,32 @@ function Translator:translate_decl(decl)
...
+
+    if tag == "ast.Exp.Unop" then
+        self:translate_exp(exp.exp)
+    elseif tag == "ast.Exp.Concat" then
+        for _, inner_exp in ipairs(exp.exps) do
+            self:translate_exp(inner_exp)
+        end
+    elseif tag == "ast.Exp.Binop" then
+        self:translate_exp(exp.lhs)
+        exp.rhs = self:translate_exp(exp.rhs)
```
Comment (hugomg) [inline pallene/translator.lua:99]: I don't think we need this `exp.rhs =`

File: pallene/translator.lua [hunk 2]
```diff
@@ -85,6 +85,32 @@ function Translator:translate_decl(decl)
...
+    elseif tag == "ast.Exp.CallFunc" then
+        self:translate_exp(exp.exp)
+        for i = 1, #exp.args do
```
Comment (hugomg) [inline pallene/translator.lua:102]: We could use ipairs here.

File: pallene/translator.lua [hunk 3]
```diff
@@ -85,16 +85,87 @@ function Translator:translate_decl(decl)
...
+        self:translate_exp(exp.rhs)
+    elseif tag == "ast.Exp.CallFunc" then
+        self:translate_exp(exp.exp)
+        for _, arg in ipairs(exp.args) do
+            self:translate_exp(arg)
+        end
+    elseif tag == "ast.Exp.Cast" then
+        self:translate_exp(exp.exp)
+        if exp.target_start_loc then
+            print('Came here')
```
Comment (hugomg) [inline pallene/translator.lua:108]: Comment

File: pallene/translator.lua [hunk 4]
```diff
@@ -85,16 +85,87 @@ function Translator:translate_decl(decl)
...
+            self:add_whitespace(exp.target_start_loc.pos, exp.target_end_loc.pos - 1)
+        end
+    elseif tag == "ast.Exp.Paren" then
+        self:translate_exp(exp.exp)
+    end
```
Comment (hugomg) [inline pallene/translator.lua:113]: That about the other kinds of expressions?

File: pallene/translator.lua [hunk 5]
```diff
@@ -85,16 +85,87 @@ function Translator:translate_decl(decl)
...
+        end
+    elseif stat._tag == "ast.Stat.While" then
+        self:translate_exp(stat.condition)
+        self:translate_stat(stat.block)
+    elseif stat._tag == "ast.Stat.Repeat" then
+        self:translate_stat(stat.block)
+        self:translate_exp(stat.condition)
+    elseif stat._tag == "ast.Stat.If" then
+        local clause = stat
+        while clause do
```
Comment (hugomg) [inline pallene/translator.lua:129]: Why are we using a while look here?

File: pallene/translator.lua [hunk 6]
```diff
@@ -85,16 +85,87 @@ function Translator:translate_decl(decl)
...
+        end
+    elseif stat._tag == "ast.Stat.Call" then
+        self:translate_exp(stat.call_exp)
+    elseif stat._tag == "ast.Stat.Return" then
+        for _, exp in ipairs(stat.exps) do
+            self:translate_exp(exp)
+        end
+    else
+        if stat._tag ~= "ast.Stat.Break" then
+            error("impossible")
```
Comment (hugomg) [inline pallene/translator.lua:168]: We would normally write this as 

```
elseif stat._tag == "ast.Stat.Break"
    -- Nothing
else
    error("impossible")
end
```

File: pallene/translator.lua [hunk 7]
```diff
@@ -141,6 +216,8 @@ end
 function translator.translate(input, prog_ast)
     local instance = Translator.new(input)
 
+    -- print(require('inspect')(prog_ast))
+
```
Comment (hugomg) [inline pallene/translator.lua:220]: Another print statement left behind. More than one today ;)

File: pallene/ast.lua [hunk 8]
```diff
@@ -70,9 +70,9 @@ declare_type("Exp", {
     Unop       = {"loc", "op", "exp"},
     Concat     = {"loc", "exps"},
     Binop      = {"loc", "lhs", "op", "rhs"},
-    Cast       = {"loc", "exp", "target"},
+    Cast       = {"exp", "target_start_loc", "target", "target_end_loc"},
```
Comment (hugomg) [inline pallene/ast.lua:73]: We might not need taget_start_loc anymore.

File: pallene/ast.lua [hunk 9]
```diff
@@ -70,9 +70,9 @@ declare_type("Exp", {
...
     Paren      = {"loc", "exp"},
-    ExtraRet   = {"loc", "call_exp", "i"}, -- See checker.lua
+ExtraRet   = {"loc", "call_exp", "i"}, -- See checker.lua
```
Comment (hugomg) [inline pallene/ast.lua:75]: Looks like you accidentally deleted some spaces.

File: pallene/translator.lua (hunk 6 above)
Comment (itssamuelrowe) [reply pallene/translator.lua:165]: Oh. I actually wrote that way and changed it. :')

File: pallene/translator.lua (hunk 7 above)
Comment (itssamuelrowe) [reply pallene/translator.lua:217]: Lots to debug :(

File: pallene/parser.lua [hunk 10]
```diff
@@ -380,7 +381,7 @@ local grammar = re.compile([[
     e10             <- (e11 {| (P op10 e11^OpExp)* |})           -> fold_binop_left
     e11             <- ({| (P unop)* |}  e12)                    -> fold_unops
     e12             <- (e13 (P op12 e11^OpExp)?)                 -> binop_right
-    e13             <- (simpleexp {| (P AS type^CastType)* |})   -> fold_casts
+    e13             <- (simpleexp {| (P AS type^CastType P)* |})   -> fold_casts
```
Comment (hugomg) [inline pallene/parser.lua:384]: Please correct the  alignment of `-> fold_casts`

**Pull request merged**